agent/stats/task.go

```diff
@@ -0,0 +1,241 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/ecscni"
+	"github.com/aws/amazon-ecs-agent/agent/stats/resolver"
+	"github.com/aws/amazon-ecs-agent/agent/utils/nswrapper"
+	"github.com/aws/amazon-ecs-agent/agent/eni/netlinkwrapper"
+	"github.com/containernetworking/plugins/pkg/ns"
+
+	dockerstats "github.com/docker/docker/api/types"
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+	netlinklib "github.com/vishvananda/netlink"
+	"math/rand"
+	"time"
+)
+
+const (
+	// linkTypeDevice defines the string that's expected to be the output of
+	// netlink.Link.Type() method for netlink.Device type.
+	linkTypeDevice = "device"
+	linkTypeVlan   = "vlan"
+	// encapTypeLoopback defines the string that's set for the link.Attrs.EncapType
+	// field for localhost devices. The EncapType field defines the link
+	// encapsulation method. For localhost, it's set to "loopback".
+	encapTypeLoopback = "loopback"
+)
+type TaskStatsInterface interface {
```

Comment:
IMO this interface is not needed and you can make netlinkClient and nsClient as fields of StatsTask and mock them in test.  GetAWSVPCNetworkStats and PopulateNIDeviceList can just be private methods of StatsTask. this way both TaskStatsInterface and TaskStatsStruct are not needed and you don't need mock for TaskStatsInterface either, so seems cleaner